pkg/volume/aws_ebs/attacher.go

```
@@ -0,0 +1,258 @@
+}
+
+func (plugin *awsElasticBlockStorePlugin) GetUniqueVolumeName(spec *volume.Spec) (string, error) {
+	volumeSource, _ := getVolumeSource(spec)
+	if volumeSource == nil {
+		return "", fmt.Errorf("Spec does not reference an EBS volume type")
+	}
+	partition := ""
+	if volumeSource.Partition != 0 {
+		partition = strconv.Itoa(int(volumeSource.Partition))
+	}
+	return fmt.Sprintf("%s/%s-%s:%v", awsElasticBlockStorePluginName, volumeSource.VolumeID, partition, volumeSource.ReadOnly), nil
+}
+
+func (attacher *awsElasticBlockStoreAttacher) Attach(spec *volume.Spec, hostName string) error {
+	volumeSource, readOnly := getVolumeSource(spec)
+	VolumeID := volumeSource.VolumeID
+
+	// Block execution until any pending detach operations for this PD have completed
+	attachDetachMutex.LockKey(VolumeID)
+	defer attachDetachMutex.UnlockKey(VolumeID)
+
+	awsCloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return err
+	}
+
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		if numRetries > 0 {
+			glog.Warningf("Retrying attach for AWS PD %q (retry count=%v).", VolumeID, numRetries)
+		}
+
+		if _, err = awsCloud.AttachDisk(VolumeID, hostName, readOnly); err != nil {
+			glog.Errorf("Error attaching PD %q: %+v", VolumeID, err)
+			time.Sleep(errorSleepDuration)
+			continue
+		}
+		return nil
+	}
+
+	return err
+}
+
+// TODO: use metadata service instead of querying cloud provider to retrieve devicePath
+func (attacher *awsElasticBlockStoreAttacher) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
+	awsCloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return "", err
+	}
+	volumeSource, _ := getVolumeSource(spec)
+	VolumeID := volumeSource.VolumeID
+	partition := ""
+	if volumeSource.Partition != 0 {
+		partition = strconv.Itoa(int(volumeSource.Partition))
+	}
+	devicePath, err := awsCloud.GetDiskPath(VolumeID)
+	if err != nil {
+		return "", err
+	}
```

@saad-ali if disk is not attached yet, it should return an error. Do you want kubelet or plugin to retry in this case?